lib/StackAnalysis/SegregateDirectStackAccesses.cpp

```diff
@@ -0,0 +1,152 @@
+  using namespace PatternMatch;
+  Value *LoadSP = nullptr;
+  bool Found = false;
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      if (auto *LI = dyn_cast<LoadInst>(&I)) {
+        if (GCBI->isSPReg(skipCasts(LI->getPointerOperand()))) {
+          LoadSP = LI;
+          Found = true;
+          break;
+        }
+      }
+    }
+
+    if (Found)
+      break;
+  }
+
+  // Context: inttoptr instructions basically inhibits all optimizations. In
+  // particular, when an integer is inttoptr'd twice with different destination
+  // type, alias analysis messes up. Hence, we need to ensure that no inttoptr
+  // exists when operating on a instruction that directly accesses the stack.
+  // Note that this problem will be addressed by opaque pointers in the future.
+  IRBuilder<> Builder(&F.getEntryBlock().front());
+  auto *I8PtrTy = Builder.getInt8PtrTy();
+  Value *SPI8Ptr = Builder
+                     .CreateLoad(I8PtrTy,
+                                 ConstantExpr::getBitCast(GCBI->spReg(),
+                                                          I8PtrTy
+                                                            ->getPointerTo()));
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      Value *Ptr = nullptr;
+
+      // Differentiate accesses and add them onto their respective bucket.
+      // Everything that is not a direct access on the stack is put onto the
+      // bucket `NotDirectStackAccesses`. Load/store that access the CSVs will
+      // have their alias info added later as well.
+      if (auto *LI = dyn_cast<LoadInst>(&I))
+        Ptr = skipCasts(LI->getPointerOperand());
+      else if (auto *SI = dyn_cast<StoreInst>(&I))
+        Ptr = skipCasts(SI->getPointerOperand());
+
+      if (Ptr) {
+        Value *LoadPtr, *LHS, *BitCast;
+        ConstantInt *Offset;
+        Builder.SetInsertPoint(&I);
+
+        // Do we have a inttoptr to a load i64, i64* LoadPtr as pointer operand
+        // of the current instruction, where LoadPtr is SP? Change it with the
+        // newly-created bitcasted load in order to prevent from using inttoptr.
+        if (match(Ptr, m_Load(m_Value(LoadPtr)))
+            && GCBI->isSPReg(skipCasts(LoadPtr))) {
+          BitCast = Builder.CreateBitCast(SPI8Ptr,
+                                          isa<LoadInst>(&I) ?
+                                            I.getOperand(0)->getType() :
+                                            I.getOperand(0)
+                                              ->getType()
+                                              ->getPointerTo());
+          I.setOperand(isa<LoadInst>(&I) ? 0 : 1, BitCast);
+
+          DirectStackAccesses.emplace_back(&I);
+
+          // Do we have a inttoptr to a add i64 LHS, -8 as pointer operand of
+          // the current instruction, where LHS is SP? Canonicalize the i2p +
+          // add into a gep whose value is bitcasted to the original type of SP.
```

Comment:
* [ ] Bad indentation, move within the next `else if`.
* [ ] Phrasing is a bit obscure. Can you try to make it clearer?
* [ ] `-8`? Are you assuming something about the architecture here? From the code it doesn't seem so.